Tests/CoreTests/UploadClientTests.swift

```diff
@@ -31,4 +31,32 @@ final class UploadClientTests: XCTestCase {
 
     self.wait(for: [uploadCompleted], timeout: 1)
   }
+
+  func testErrorResponsesAreThrowAsErrors() async throws {
+    let errorMessage = "Something went wrong"
+    let errorResponse = UploadFailureResponse(message: errorMessage)
+
+    let uploadClient = UploadClient.live(
+      api: .init(
+        decoder: JSONDecoder(),
+        request: { _ in
+          (try JSONEncoder().encode(errorResponse), URLResponse(url: URL(string: "test")!, mimeType: nil, expectedContentLength: 0, textEncodingName: nil))
+        }
+        )
+      )
+    do {
+      try await uploadClient.upload(trace: Trace(id: "id", history: .init(section: "section")))
+    } catch {
+      guard let error = error as? UploadClient.UploadError,
+            case let .error(message) = error else {
+        XCTFail("Thrown error should be UploadError type")
+        return
+      }
+      
+      XCTAssertEqual(message, errorMessage)
+      return
+    }
+
+    XCTFail("Did not throw error with error response")
+  }
```

Comment:
Thoughts on something like this? It's just a bit restructured and I thought we could test against `localizedDescription` if we conform `UploadError` to `LocalizedError` because that is what is passed to the logger.  

```suggestion
  func testFailureResponseThrowsAsError() async throws {
    let data = try JSONEncoder().encode(UploadFailureResponse(message: "Something went wrong"))
    let api = ApiClient { _ in (data, .success()) }
    let uploadClient = UploadClient.live(api: api)
    let trace = Trace(id: "id", history: .init(section: "section"))

    let task = Task { try await uploadClient.upload(trace: trace) }
    let result = await task.result
    
    XCTAssertThrowsError(try result.get()) { error in
      XCTAssertTrue(error is UploadClient.UploadError)
      XCTAssertEqual(error.localizedDescription, "Something went wrong")
    }
  }
```

Reply:
Way nicer

Reply:
Perhaps we should rename the extension function on URLResponse from `.success()` to `stub()` since in this case we're using it for a failure case

Reply:
> from .success() to .stub()

Yep. great idea :)